### PR TITLE
Decode the public input from compliance public memory

### PR DESCRIPTION
**Pull request by CarloModicaPortfolio**

### CODE REVIEW

File: native/cairo_prover/src/lib.rs [hunk 1]
```diff
@@ -105,4 +106,34 @@ fn cairo_verify(proof: Vec<u8>, public_input: Vec<u8>) -> bool {
     verify_cairo_proof(&proof, &pub_inputs, &proof_options)
 }
 
-rustler::init!("Elixir.Cairo.CairoProver", [cairo_prove, cairo_verify]);
+#[rustler::nif(schedule = "DirtyCpu")]
```
Comment (XuyangSong) [inline native/cairo_prover/src/lib.rs]: It should be very fast. We may not need the "DirtyCpu" here.

File: native/cairo_prover/src/lib.rs [hunk 2]
```diff
@@ -105,4 +106,34 @@ fn cairo_verify(proof: Vec<u8>, public_input: Vec<u8>) -> bool {
...
+            eprintln!("Error: 'Output' segment not found in memory_segments");
+            return Ok(vec![]);
+        }
+    };
+
+    let begin_addr :u64 = output_segments.begin_addr.try_into().unwrap();
+    let stop_addr :u64 = output_segments.stop_ptr.try_into().unwrap();
+
+    let mut output_values = Vec::new();
+    for addr in begin_addr..=stop_addr-1 {
```
Comment (XuyangSong) [inline native/cairo_prover/src/lib.rs]:
```suggestion
    for addr in begin_addr..stop_addr {
```
It seems identical but more concise?

File: native/cairo_prover/src/lib.rs [hunk 3]
```diff
@@ -105,4 +106,34 @@ fn cairo_verify(proof: Vec<u8>, public_input: Vec<u8>) -> bool {
...
+        // Convert addr to FieldElement (assuming this is the correct way to create a FieldElement from an address)
+        let addr_field_element = Felt252::from(addr);
+
+        if let Some(value) = pub_inputs.public_memory.get(&addr_field_element) {
+            output_values.push(value.clone().to_bytes_le().to_vec());
+        } else {
+            eprintln!("Error: Address {:?} not found in public memory", addr_field_element);
+        }
+    }
+    println!("{:?}", output_values.clone());
```
Comment (XuyangSong) [inline native/cairo_prover/src/lib.rs]: the print here is for test?

File: test/cairo_compliance_test.exs [hunk 4]
```diff
@@ -16,6 +16,8 @@ defmodule CairoComplianceTest do
 
     # Prove and verify
     {proof, public_input} = Cairo.prove(trace, memory, public_inputs)
-    assert true = Cairo.verify(proof, public_input)
+    # assert true = Cairo.verify(proof, public_input)
```
Comment (XuyangSong) [inline test/cairo_compliance_test.exs]: Do we want to skip the verify in the test?

**Review by XuyangSong — CHANGES_REQUESTED** [review]

File: native/cairo_prover/src/lib.rs (hunk 3 above)
Comment (camofu) [reply native/cairo_prover/src/lib.rs]: Yes, it was. I forgot to take it out, will fix it.

File: test/cairo_compliance_test.exs (hunk 4 above)
Comment (camofu) [reply test/cairo_compliance_test.exs]: Same as above

**Pull request merged**